src/Microsoft.AspNetCore.Razor.Runtime/TagHelpers/TagHelperOutput.cs

```diff
@@ -183,17 +187,26 @@ public bool IsContentModified
         /// a <c>Microsoft.AspNetCore.Mvc.Rendering.HtmlString</c> instance. MVC converts most other types to a
         /// <see cref="string"/>, then HTML encodes the result.
         /// </remarks>
-        public TagHelperAttributeList Attributes
+        public TagHelperAttributeList Attributes { get; }
+
+        /// <summary>
+        /// Clears the <see cref="TagHelperOutput"/> and updates its state with the provided values.
+        /// </summary>
+        /// <param name="tagName">The tag name to use.</param>
+        /// <param name="tagMode">The <see cref="TagMode"/> to use.</param>
+        public void Reinitialize(string tagName, TagMode tagMode)
         {
-            get
-            {
-                if (_attributes == null)
-                {
-                    _attributes = new TagHelperAttributeList();
-                }
+            TagName = tagName;
+            TagMode = tagMode;
+            Attributes.Clear();
 
-                return _attributes;
-            }
+            _preElement?.Reset();
```

Comment:
Setter still needed unless we reorganized the code gen semi-drastically to not set output.Content (it does it today to prevent IsModified from being set).